docs/proposals/workflow.md

```diff
@@ -0,0 +1,134 @@
+
+## Implementation
+
+The basic idea consists in adding a label selector to the current Job API object. The new selector will determine the parent jobs.  The parent jobs are the jobs current job will depend on. The current job will be scheduled once all the parent jobs will run to completion.
+The strongest point in favor of this approach is the ability to re-use the current `job` implementation. Using a label selector the current `job` will simply become a `graph` of jobs: all the features/controllers that will be built on top of `job` like for example [#11980](https://github.com/kubernetes/kubernetes/issues/11980) and [#17242](https://github.com/kubernetes/kubernetes/issues/17242) will be automatically extended to support workflows. Not sure about https://github.com/kubernetes/kubernetes/issues/16845 and https://github.com/kubernetes/kubernetes/issues/14188.
+A similar approach is implemented in Chronos for [dependent jobs](https://mesos.github.io/chronos/docs/api.html#adding-a-dependent-job).
+
+### API Object
+
+To implement _workflows_ the `Job` API should be modified:
+
+```go
+// JobsSelector it's just an alias for PodSelector
+type JobSelector PodSelector
+
+// Job represents the configuration of a single job.
+type Job struct {
+    unversioned.TypeMeta `json:",inline"`
+    // Standard object's metadata.
+
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+    api.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is a structure defining the expected behavior of a job.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+    Spec JobSpec `json:"spec,omitempty"`
+
+    // Status is a structure describing current status of a job.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+    Status JobStatus `json:"status,omitempty"`
+
+    // JobSelector to detect parents jobs
+    ParentSelector *JobSelector `json:"parentSelector, omitempty"`
+}
+```
+
+#### JobSpec and JobStatus
+
+No modifications
+
+#### Parent JobSelector
+
+The `.job.parentselector` is a label query over a set of jobs.
+If all selected jobs are completed the `job` will be started immediately.
+If `.job.parentselector` is absent the `job` will be started immediately.
+Since labels are forwarded directly from `job` to pods, users should not create pods whose labels match this selector, either directly,
+via another Job, or via another controller (for example Replication controller), [see](https://github.com/kubernetes/kubernetes/issues/14961).
+
+#### Some special cases
+
+Jobs terminated via [#17244](https://github.com/kubernetes/kubernetes/issues/17244)  (if implemented) will be considered _completed_ so will trigger the _child_ jobs. Since it's very unlikely that the user would keep the _child_ jobs alive we may add a field to cascade termination to _child_ jobs if any to the current proposals made by @pmorie and @bgrant0607.
```

Comment:
Actually not necessarily. Both #17244 and #17242 will introduce new condition - fail. In that case we should be default not to run child jobs. The question rather is - if we ever want to allow them? I can't think of any reasonable use-cases, but we could provide an option for that.